### PR TITLE
[C-2678] Add Stems and Source Files Modal

### PR DESCRIPTION
**Pull request by amendelsohn**
### Description

- Stems and Source Files modal
- Ported stems file uploader component

Also:
- fixes input value association for deep fields like `fieldVisibility.remixes`

Todo:
- link the toggles together so you can only enable follow required if the download is enabled

image

### Feature Flags ###

`UPLOAD_REDESIGN_ENABLED`

### CODE REVIEW

File: packages/web/src/pages/upload-page/fields/ReleaseDateModalForm.tsx [hunk 1]
```diff
@@ -28,15 +28,14 @@ export const ReleaseDateModalForm = () => {
...
 
-  const initialValues = useMemo(
-    () => ({
-      [RELEASE_DATE]: value ?? null
-    }),
-    [value]
-  )
+  const initialValues = useMemo(() => {
+    const initialValues = {}
+    set(initialValues, RELEASE_DATE, value)
```
Comment (dylanjeffers) [inline packages/web/src/pages/upload-page/fields/ReleaseDateModalForm.tsx:33]: just double checking the reason we need this is cause RELEASE_DATE could be a `a.b` type key?

File: packages/web/src/pages/upload-page/fields/SourceFilesModalForm.module.css [hunk 2]
```diff
@@ -0,0 +1,41 @@
+.header {
```
Comment (dylanjeffers) [inline packages/web/src/pages/upload-page/fields/SourceFilesModalForm.module.css:1]: i think default is 2 spaces, run `npm run stylelint` to find out. might want to update vscode for lint on save here, can help with that :)

File: packages/web/src/pages/upload-page/fields/SourceFilesModalForm.tsx [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+  const initialValues = useMemo(() => {
+    const initialValues = {}
+    set(initialValues, ALLOW_DOWNLOAD, allowDownloadValue)
+    set(initialValues, FOLLOWER_GATED, followerGatedValue)
+    set(initialValues, STEMS, stemsValue ?? [])
+    return initialValues as SourceFilesFormValues
+  }, [allowDownloadValue, followerGatedValue, stemsValue])
+
+  const onSubmit = (values: SourceFilesFormValues) => {
```
Comment (dylanjeffers) [inline packages/web/src/pages/upload-page/fields/SourceFilesModalForm.tsx:65]: if this is new every time there is no point in memoizing the initial values, there is a chance we can just not worry about the memo, up to you

File: packages/web/src/pages/upload-page/fields/SourceFilesModalForm.tsx [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+  }
+
+  const preview = (
+    <div className={styles.preview}>
+      <div className={styles.header}>
+        <label className={styles.title}>{messages.title}</label>
+      </div>
+      <div className={styles.description}>{messages.description}</div>
+    </div>
+  )
```
Comment (dylanjeffers) [inline packages/web/src/pages/upload-page/fields/SourceFilesModalForm.tsx:78]: such a sweet file!!

File: packages/web/src/pages/upload-page/fields/SourceFilesView.tsx [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+  onDeleteStem: (index: number) => void
+}
+
+export const SourceFilesView = ({
+  onAddStems,
+  stems,
+  onSelectCategory,
+  onDeleteStem
+}: SourceFilesViewProps) => {
+  const renderCurrentStems = () => {
```
Comment (dylanjeffers) [inline packages/web/src/pages/upload-page/fields/SourceFilesView.tsx:37]: ul, li would be ideal :)

File: packages/web/src/pages/upload-page/fields/SourceFilesView.tsx [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+  StemCategory.OTHER
+]
+
+type StemRowProps = {
+  stem: StemUpload
+  didSelectCategory: (category: StemCategory) => void
+  onDelete: () => void
+}
+
+const StemRow = ({
```
Comment (dylanjeffers) [inline packages/web/src/pages/upload-page/fields/SourceFilesView.tsx:115]: row is fine, but also "list-item" i typically use for lists, just a headsup

File: packages/web/src/pages/upload-page/fields/SourceFilesView.tsx [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+
+  let stemIndex = dropdownRows.findIndex((r) => r === category)
+  if (stemIndex === -1) {
+    console.error(`Couldn't find stem row for category: ${category}`)
+    stemIndex = 0
+  }
+
+  const renderDeleteButton = () => {
+    return (
+      <div className={styles.deleteButton}>
```
Comment (dylanjeffers) [inline packages/web/src/pages/upload-page/fields/SourceFilesView.tsx:133]: span?

File: packages/web/src/pages/upload-page/fields/SourceFilesView.tsx (hunk 6 above)
Comment (dylanjeffers) [reply packages/web/src/pages/upload-page/fields/SourceFilesView.tsx:115]: i assume a lot of this is from the old inputs anyway, so up to you on amount of refactoring

**Review by dylanjeffers — APPROVED** [review]
so clutch all around great job. some of the form logic is a little hard to follow, with the sets an the !!.some() but overall excellent

File: packages/web/src/pages/upload-page/fields/SourceFilesModalForm.module.css (hunk 2 above)
Comment (amendelsohn) [reply packages/web/src/pages/upload-page/fields/SourceFilesModalForm.module.css:1]: Ah true. Thanks!

As in we have setup for this and mine is broken or like we can add it?

Comment (dylanjeffers) [reply packages/web/src/pages/upload-page/fields/SourceFilesModalForm.module.css:1]: I think your setup is maybe broken and can help you set that up. But yeah to fix,run stylelint commands

**Comment by audius-infra** [comment]
Preview this change https://demo.audius.co/amendel-upload-stems-modal

File: packages/web/src/pages/upload-page/fields/SourceFilesModalForm.module.css (hunk 2 above)
Comment (amendelsohn) [reply packages/web/src/pages/upload-page/fields/SourceFilesModalForm.module.css:1]: I think I fixed it. Added the stylelint plugin and it's running on save. Thanks!

File: packages/web/src/pages/upload-page/fields/ReleaseDateModalForm.tsx (hunk 1 above)
Comment (amendelsohn) [reply packages/web/src/pages/upload-page/fields/ReleaseDateModalForm.tsx:37]: yes, I'm using it because useField supports lodash-style nested field names
https://formik.org/docs/api/field#name

**Comment by audius-infra** [comment]
Preview this change https://demo.audius.co/amendel-upload-stems-modal
